DSEGOG-95 search by shot number

sam-glendenning opened the pull request:
## Description
Allows the user to search by shot number through a popup window containing min and max values, as per the wireframes. I've refactored the functionality behind searching as a whole so that the existing date-time search functionality works in tandem with it. As a result, there are a couple of breaking changes:

- search is launched when the user presses the big shiny SEARCH button. This replaces previous behaviour in the date-time pickers which updated the search parameters whenever the user updated their values. Whatever values are currently in the search fields will be used in the new search query and the table won't be updated until the search button is pressed again with new search entries
- a need to tie the search parameters in with the existing filters work mean all the search and filter conditions passed to the backend are wrapped in one big `$and` statement. As far as I can tell, this works as expected but please test this yourself
- The wireframes seem to indicate that min and max fields for searching by a shot number range are always present but I may have this wrong. Therefore, as it stands, a min and max parameter for shot number of 0 and 99999999 respectively are included with every query (unless the user changes them). This has complicated queries and unit tests. It may be that we decide to make these optional, in which case I'll have to change the design a bit

## Testing instructions
Test searching _and_ filtering on various fields and check the data being displayed matches what you expect.

- [x] Review code
- [x] Check Actions build
- [x] Review changes to test coverage

## Agile board tracking
Closes DSEGOG-95

Commit:
DSEGOG-95 adding search bar and calendar icon for date pickers

Adding a search folder for all components of the search bar

Moving date search to it and adding a calendar icon, as per the wireframes

Commit:
DSEGOG-95 renaming datetimeinputbox

To dateTimeSearch, as this is now more appropriate as a subcomponent of searching

Commit:
DSEGOG-95 reducing date input width

Commit:
DSEGOG-95 adding timeframe search template

Commit:
Merge branch 'develop' into DSEGOG-95-search-by-shot-id

Commit:
DSGOG-95 adding other search components

Commit:
DSEGOG-95 adding shot number dialog

Includes two number input fields

Commit:
DSEGOG-95 updating search method with search button

The search button is now the sole method of launching a search via the input parameters. This means that the searchSlice has been modified such that its state only updates when the user clicks the search button

Commit:
DSEGOG-95 adding shotnumRange to query

This needed a refactor to how conditions are passed to the backend query

Commit:
DSEGOG-95 fixing timestamp search bug

For some reason, the timestamp search params were being added to the query when none were present but only when a shotnum range outside the default was present. Why this happens, I don't know, but I've added a more stringent check to prevent it from occurring

Commit:
DSEGOG-95 fixing some existing tests

Also adding ability to prevent empty query params object being sent as part of query. This ensures existing tests pass and prevents queries from being laden down by unnecessary objects

Commit:
DSEGOG-95 shotnum range is now required, adding unit test

Making the min and max fields of shotnumRange required fields again. This makes testing easier and brings the component back in line with how the wireframes depict it, so better mimics how the user will interact with it

Commit:
DSEGOG-95 adding red border when invalid range

Border colour of shot num range changes to red when there is an error in its input (e.g. invalid range). This can be used to highlight an issue when the input is closed and the user may be wondering what's wrong with the search

Commit:
Adding @babel/runtime dependency

May help fix a babel/runtime dependency issue me and Louise have spotted occasionally

Commit:
DSEGOG-95 adding shot number search tests

Also improving method by which we change the method of typing in shot number range input values to ensure the correct values are being entered

Commit:
DSEGOG-95 expanding searchBar tests

Adding tests for filling in date-time fields

Commit:
DSEGOG-95 adding another searchBar test

Commit:
DSEGOG-95 adding e2e tests for searching

Also amending existing tests to account for new conditions param for records and count endpoints

Commit:
Merge branch 'develop' into DSEGOG-95-search-by-shot-id

Commit:
DSEGOG-95 fixing unit tests

Commit:
DSEGOG-95 increasing SLOPPY_CLICK_THRESHOLD

It could be because I've slightly amended the  dataHeader widths but the drag and drop functionality in e2e tests was failing. I've increased the threshold for clicking and dragging to try and fix this

Commit:
Merge branch 'develop' into DSEGOG-95-search-by-shot-id

Commit:
Merge branch 'develop' into DSEGOG-95-search-by-shot-id

Commit:
Remove old @babel/runtime dependency

sam-glendenning commented:
Following a conversation, we have decided to make the min and max fields optional. They can be greyed-out "default" values in the input boxes but not selected unless the user edits them

Commit:
DSEGOG-95 shotnum range now optional

Min and max fields are not populated by default

Commit:
DSEGOG-95 updating tests to account for optional shotnum range

Also adding additional e2e test for red error highlight on bad search params

sam-glendenning commented:
@louise-davies ready for a further review with the above changes implemented (optional shotnumber range)

louise-davies commented on src/search/components/dateTime.component.tsx:
```diff
@@ -20,100 +19,95 @@ export interface UpdateFilterParams {
                 updateFilter({
                   date: date as Date,
-                  prevDate: initialFromDate,
+                  prevDate: fromDate,
```
This reintroduces the bug we had previously, where "half" selecting a date doesn't work. We need to use `receivedFromDate`/`receivedToDate` in `prevDate`. See gif

![OG-timestamp-search-bug](https://user-images.githubusercontent.com/22525471/203573257-03165a4a-4320-4f1a-8369-0ad45e94a671.gif)

Also, as an aside, the box for the from date/to date is overly large on my monitor, with a lot of deadspace on the right

sam-glendenning replied:
Thanks for pointing that out. I'm annoyed I let that back in there. And the deadspace issue - a constant reminder I need to work on my CSS skills

louise-davies reviewed (changes requested):
The shot number functionality works well - just need to fix the issue with the date/time pickers

Commit:
Merge branch 'develop' into DSEGOG-95-search-by-shot-id

Commit:
DSEGOG-95 fixing date-time "half-selection" bug

Commit:
DSEGOG-95 fixing CSS overflow and width issues

This should make the search bar look nice at a variety of zoom levels

Commit:
DSEGOG-95 reverting increase to SLOPPY_CLICK_THRESHOLD

Whatever was making this fail before, it's now gone. This ensures the tests to move the columns in the table pass
